### ISSUE TITLE
关于特征和label的编号问题

### ISSUE THREAD

**Issue by shuizhonghaitong** [liu-nlper/NER-LSTM-CRF#21]
您好，看您的代码，特征的编号是从2开始，label的编号是从1开始，nb_classes是voc[-1]+1，所以总共的标签数是7的时候，nb_classes是9，这里不太懂，nb_classes实际上不应该仍然是7吗，我不太清楚这里的逻辑，求讲解，谢谢！

**Comment by liu-nlper** [comment]
您好：

  不好意思回复这么晚，最近在忙毕业论文的事情。

  感谢指出问题，你说的那个问题是我写错了，label是从1开始编号，0留作padding值的标签，所以`nb_classes`应该是7+1=8。

  另外，修正了`use_crf`设置为false的问题。

祝好

2018-03-23 9:58 GMT+08:00 liuyutong <notifications@github.com>:

**Comment by shuizhonghaitong** [comment]
非常感谢您的作答。不好意思，我比较业余，还有些不太懂。为什么nb_classes和标签数不是相同的，padding值是什么意思？还有，特征为什么是从2开始编号？

**Comment by liu-nlper** [comment]
@shuizhonghaitong 
1. 由于每个句子的长度不一致，所以在加载数据时需要将他们补全（padding）为固定的长度，例如句子实际长度为5，但是max_len设置的是10，则sentence = [1,2,3,4,4,0,0,0,0,0]，这里1,2,3,4是lable的标签，0即为padding值的标签。
2. 为什么特征从2开始编号：同样地，数字0是留作padding值的编号，数字1是未登录词的编号，故从2开始编号。

**Comment by shuizhonghaitong** [comment]
还有一个问题，为什么这里需要考虑padding值的编号，给tf.nn.bidirectional_dynamic_rnn提供了sequence_length的参数后，应该能自动去掉那些padding了？

**Comment by liu-nlper** [comment]
传入tf.nn.bidirectional_dynamic_rnn的数据可能padding值设为任意数值都行，
但是这样就没办法从输入的sentence计算其实际长度，例如sentence =
[1,2,3,4,1,0,0,0,0]，那么该sentence的实际长度就可以通过tf.reduce_sum(tf.sign(sentence))计算出；若padding值为1，那么sentence=[1,2,3,4,1,1,1,1,1,1]就没有办法算出。
另外，若使用CNN层，那么padding值具有特别的含义，最好不要与其他单词是同一个id。

**Comment by shuizhonghaitong** [comment]
终于明白了，真的太感谢了，您回答的这么详细这么清晰：）
